lib/src/Uncertainty/Process/DiscreteMarkovChain.cxx

```diff
@@ -0,0 +1,243 @@
+TimeSeries DiscreteMarkovChain::getFuture(const UnsignedInteger stepNumber) const
+{
+  // TimeGrid of the process
+  RegularGrid timeGrid;
+  try
+  {
+    timeGrid = getTimeGrid();
+  }
+  catch (...)
+  {
+    throw InternalException(HERE) << "Error: can extend the realization of a process only if defined on a regular grid.";
+  }
+  if (stepNumber == 0) throw InvalidArgumentException(HERE) << "Error: the number of future steps must be positive.";
+  
+  // TimeGrid associated with the possible future
+  const Scalar timeStep = RegularGrid(mesh_).getStep();
+  const RegularGrid futurTimeGrid(RegularGrid(mesh_).getEnd(), timeStep, stepNumber);
+  const UnsignedInteger size = futurTimeGrid.getVerticesNumber();
+  
+  // Definition of the states of the system
+  const UnsignedInteger dimension = transitionMatrix_.getNbRows();
+  Sample states(dimension, 1);
+  for (UnsignedInteger i = 0; i < dimension; i++)
+  {
+    states(i, 0) = i;
+  }
+
+  // Loop on mesh points
+  UnsignedInteger previous(currentState_);
+  Sample data(size, 1);
+  data(0, 0) = currentState_;
+  for (UnsignedInteger i = 1; i < size; i++)
+  {
+    // Definition of the distribution
+    Point weights(dimension, 0.0);
+    for (UnsignedInteger col=0; col < dimension; col++)
+    {
+      weights[col] = transitionMatrix_(previous, col);
+    }
+    UserDefined dist = UserDefined(states, weights);
+    
+    // Get a realization of the distribution
+    data[i] = dist.getRealization();
+    previous = data(i, 0);
+  }
+  return TimeSeries(futurTimeGrid, data);
+}
+
+/* Transition matrix accessors */
+SquareMatrix DiscreteMarkovChain::getTransitionMatrix() const
+{
+  return transitionMatrix_;
+}
+
+void DiscreteMarkovChain::setTransitionMatrix(const SquareMatrix & transitionMatrix)
+{ 
+    // Check transition matrix dimension consistency with current state
+    if (transitionMatrix.getNbRows() <= currentState_) 
+    {
+      LOGWARN(OSS()<< "Warning: the given transition matrix has a dimension = " << transitionMatrix.getNbRows() << " incompatible with the current state = " << currentState_<<". Current state is set to 0.");
+      currentState_ = 0;
+    }
+    
+    // Check transition matrix dimension consistency with origin
+    if (transitionMatrix.getNbRows() <= origin_) 
+    {
+      LOGWARN(OSS()<< "Warning: the given transition matrix has a dimension = " << transitionMatrix.getNbRows() << " incompatible with the origin state = " << origin_<<". Origin state is set to 0.");
+      setOrigin(0);
+    }
+    
+    // Check if transition matrix is stochastic
+    Point ones(transitionMatrix.getNbRows(), 1.0);
+    Point sums(transitionMatrix * ones);
+    for (UnsignedInteger i = 0; i < transitionMatrix.getNbRows(); i++)
+    {
+        if (fabs(sums[i] - 1.0) > SpecFunc::Precision) throw InvalidArgumentException(HERE) << "Error: the given transition matrix is not stochastic. Sum of line " << i+1 << " is equal to " << sums[i];
```

Comment:
std::abs

Reply:
you could also normalize the probabilities as there is a 100% chance of failure for this test when the dimension is large. At least the threshold should be user-defined eg using ResourceMap.